### PR TITLE
benchtest: break out of infinite loop and fix goroutine leak

### PR DESCRIPTION
**Pull request by kruskall**
## Motivation/summary

Another goroutine leak in benchtest:

```
--- FAIL: TestWatchMetricNonBlocking (1.51s)
    leaks.go:78: found unexpected goroutines:
        [Goroutine 29 in state chan send, with github.com/elastic/apm-server/systemtest/benchtest/expvar.run.func1 on top of the stack:
        goroutine 29 [chan send]:
        github.com/elastic/apm-server/systemtest/benchtest/expvar.run.func1()
                /apm-server/systemtest/benchtest/expvar/metrics.go:251 +0x105
        created by github.com/elastic/apm-server/systemtest/benchtest/expvar.run
                /apm-server/systemtest/benchtest/expvar/metrics.go:231 +0x110
        [originating from goroutine 35]:
        github.com/elastic/apm-server/systemtest/benchtest/expvar.run(...)
                /apm-server/systemtest/benchtest/expvar/metrics.go:255 +0x110
        github.com/elastic/apm-server/systemtest/benchtest/expvar.(*Collector).start(...)
                /apm-server/systemtest/benchtest/expvar/metrics.go:210 +0xf6
        github.com/elastic/apm-server/systemtest/benchtest/expvar.TestAggregate(...)
                /apm-server/systemtest/benchtest/expvar/metrics_test.go:52 +0x1aa
        github.com/elastic/apm-server/systemtest/benchtest/expvar.StartNewCollector(...)
                /apm-server/systemtest/benchtest/expvar/metrics.go:80 +0x10b
        testing.tRunner(...)
                /usr/lib/go/src/testing/testing.go:1579 +0x10b
        created by testing.(*T).Run
                /usr/lib/go/src/testing/testing.go:1629 +0x3ea
        [originating from goroutine 1]:
        testing.(*T).Run(...)
                /usr/lib/go/src/testing/testing.go:1630 +0x3ea
        testing.runTests.func1(...)
                /usr/lib/go/src/testing/testing.go:2035 +0x45
        testing.tRunner(...)
                /usr/lib/go/src/testing/testing.go:1579 +0x10b
        testing.runTests(...)
                /usr/lib/go/src/testing/testing.go:2040 +0x489
        testing.(*M).Run(...)
                /usr/lib/go/src/testing/testing.go:1906 +0x63a
        main.main(...)
                _testmain.go:57 +0x1aa
        
         Goroutine 277 in state chan send, with github.com/elastic/apm-server/systemtest/benchtest/expvar.run.func1 on top of the stack:
        goroutine 277 [chan send]:
        github.com/elastic/apm-server/systemtest/benchtest/expvar.run.func1()
                /apm-server/systemtest/benchtest/expvar/metrics.go:248 +0x1fb
        created by github.com/elastic/apm-server/systemtest/benchtest/expvar.run
                /apm-server/systemtest/benchtest/expvar/metrics.go:231 +0x110
        [originating from goroutine 202]:
        github.com/elastic/apm-server/systemtest/benchtest/expvar.run(...)
                /apm-server/systemtest/benchtest/expvar/metrics.go:255 +0x110
        github.com/elastic/apm-server/systemtest/benchtest/expvar.(*Collector).start(...)
                /apm-server/systemtest/benchtest/expvar/metrics.go:210 +0xf6
        github.com/elastic/apm-server/systemtest/benchtest/expvar.TestWatchMetricNonBlocking(...)
                /apm-server/systemtest/benchtest/expvar/metrics_test.go:119 +0x1aa
        github.com/elastic/apm-server/systemtest/benchtest/expvar.StartNewCollector(...)
                /apm-server/systemtest/benchtest/expvar/metrics.go:80 +0x10b
        testing.tRunner(...)
                /usr/lib/go/src/testing/testing.go:1579 +0x10b
        created by testing.(*T).Run
                /usr/lib/go/src/testing/testing.go:1629 +0x3ea
        [originating from goroutine 1]:
        testing.(*T).Run(...)
                /usr/lib/go/src/testing/testing.go:1630 +0x3ea
        testing.runTests.func1(...)
                /usr/lib/go/src/testing/testing.go:2035 +0x45
        testing.tRunner(...)
                /usr/lib/go/src/testing/testing.go:1579 +0x10b
        testing.runTests(...)
                /usr/lib/go/src/testing/testing.go:2040 +0x489
        testing.(*M).Run(...)
                /usr/lib/go/src/testing/testing.go:1906 +0x63a
        main.main(...)
                _testmain.go:57 +0x1aa
        ]
```

Once the ctx is done the return statement will break out of the `select` and the parent for loop will run forever repeating the same loop.
Another side effect is that `defer c.cleanup()` is unreachable code as the goroutine will never finish.
Finally, sending to `errChan` or `outChan` could potentially block forever causing another goroutine leak if the function reading from the channels has already completed.

## Checklist

- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)
- [ ] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

### CODE REVIEW

**Review by axw — COMMENTED** [review]
Are the `break outer` statements necessary? As far as I can see, they're essentially equivalent to the existing `return` statements, since the loops are at the ends the functions.

I think the only changes necessary are adding a `select` around the channel sends, so they don't block indefinitely.

**Comment by kruskall** [comment]
> Are the `break outer` statements necessary? As far as I can see, they're essentially equivalent to the existing `return` statements, since the loops are at the ends the functions.
> 
> I think the only changes necessary are adding a `select` around the channel sends, so they don't block indefinitely.

I think I'm missing something, the return applies to the innermost block (select) so the loop will go on forever.
Consider the following:

```
package main

import (
	"fmt"
	"time"
)

func main() {
	defer fmt.Println("Cleanup")
	for {
		select {
		case <-time.After(1 * time.Second):
			return
		default:
			fmt.Println("select")
		}
	}
}
```

The above code will print `select` forever and `Cleanup` will never be executed.

**Comment by axw** [comment]
> I think I'm missing something, the return applies to the innermost block (select) so the loop will go on forever.

Nope. Per [the spec](https://go.dev/ref/spec#Return_statements):

> A "return" statement in a function F terminates the execution of F, and optionally provides one or more result values. Any functions [deferred](https://go.dev/ref/spec#Defer_statements) by F are executed before F returns to its caller.

So to be clear, `return` terminates the enclosing function regardless of where in the function the statement is.

The program in https://github.com/elastic/apm-server/pull/10298#issuecomment-1437867766 is not equivalent to the code you changed in the PR, due of the `default` case. Try this instead:

```go
package main

import (
	"fmt"
	"time"
)

func main() {
	defer fmt.Println("Cleanup")
	t := time.After(1 * time.Second)
	for {
		select {
		case <-t:
			return
		default:
			fmt.Println("select")
		}
	}
}
```

**Comment by kruskall** [comment]
:facepalm: 

Updated!

File: systemtest/benchtest/expvar/metrics.go [hunk 1]
```diff
@@ -245,10 +246,18 @@ func run(ctx context.Context, serverURL string, period time.Duration) (<-chan ex
 				err := queryExpvar(ctxWithTimeout, &e, serverURL)
 				cancel()
 				if err != nil {
-					errChan <- err
+					select {
+					case errChan <- err:
+					case <-ctx.Done():
+						return
```
Comment (axw) [inline systemtest/benchtest/expvar/metrics.go:252]:
```suggestion
```

**Review by axw — APPROVED** [review]
thanks :)

**Pull request merged**